nameko/containers.py

```diff
@@ -353,41 +348,55 @@ def spawn_worker(self, entrypoint, args, kwargs,
             self, service, entrypoint, args, kwargs, data=context_data)
 
         _log.debug('spawning %s', worker_ctx)
-        gt = self._worker_pool.spawn(self._run_worker, worker_ctx,
-                                     handle_result)
-        self._active_threads[gt] = entrypoint
-        gt.link(self._handle_thread_exited)
+        gt = self._worker_pool.spawn(
+            self._run_worker, worker_ctx, handle_result
+        )
+        self._worker_threads[worker_ctx] = gt
+        self.worker_created(worker_ctx)
+        gt.link(self._handle_worker_thread_exited, worker_ctx)
         return worker_ctx
 
-    def spawn_managed_thread(self, run_method, protected=False):
-        """ Spawn a managed thread to run ``run_method``.
-
-        Threads can be marked as ``protected``, which means the container will
-        not forcibly kill them until after all extensions have been killed.
-        Extensions that require a managed thread to complete their kill
-        procedure should ensure to mark them as ``protected``. For example,
-        :class:`nameko.messaging.QueueConsumer` cleanly closes connections
-        to the AMQP broker even when killed; the thread that holds that
-        connection is ``protected`` so it isn't stopped until after the
-        QueueConsumer returns from its kill.
+    def spawn_managed_thread(self, fn, extension=None, protected=None):
+        """ Spawn a managed thread to run ``fn`` on behalf of `extension`.
 
-        Any uncaught errors inside ``run_method`` cause the container to be
-        killed.
+        Any uncaught errors inside ``fn`` cause the container to be killed.
 
         It is the caller's responsibility to terminate their spawned threads.
         Threads are killed automatically if they are still running after
         all extensions are stopped during :meth:`ServiceContainer.stop`.
 
         Extensions should delegate all thread spawning to the container.
         """
-        gt = eventlet.spawn(run_method)
-        if not protected:
-            self._active_threads[gt] = MANAGED_THREAD
-        else:
-            self._protected_threads.add(gt)
-        gt.link(self._handle_thread_exited)
+        if not isinstance(extension, Extension):
+            warnings.warn(
+                "The signature of `spawn_managed_thread` has changed. "
+                "The `protected` kwarg is now deprecated, and extensions "
+                "should pass themselves as the second argument for better "
+                "logging. See :meth:`nameko.containers.ServiceContainer."
```

Comment:
passing `self` seems a bit ugly. also, if this is just for logging, do we want to be more explicit and call it out as that? an extension using multiple managed threads could then request them with different info, e.g. `MyExt.method_foo`